File: hw5/SplitFileClient.go
/*
SplitFileClient.go
20190532 sang yun lee
*/
package main

import (
	"bufio"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strconv"
)

func main() {
	// argument handling
	if len(os.Args) != 3 {
		fmt.Println("Invalid argument.")
		os.Exit(1)
	}
	// get argument from system args
	commandName := os.Args[1]
	fileName := os.Args[2]

	// server Info hardcoding
	firstServerName := "127.0.0.1"
	firstServerPort := "20532"
	secondServerName := "127.0.0.1"
	secondServerPort := "20532"

	// put command case
	if commandName == "put" {
		sendFile(fileName, firstServerName, firstServerPort, 0)
		sendFile(fileName, secondServerName, secondServerPort, 1)
	} else if os.Args[1] == "get" { // get command case
		receiveFile(fileName, firstServerName, firstServerPort, 0)
		receiveFile(fileName, secondServerName, secondServerPort, 1)

		// get file extension from fileName
		fileExtension := filepath.Ext(fileName)
		// get file name without extension from fileName
		fileName = fileName[0 : len(fileName)-len(fileExtension)]
		// file create
		file, err := os.Create(fileName + "-merged" + fileExtension)
		if err != nil {
			fmt.Println("fail to create file:", err)
			return
		}
		tmpFileName1 := fmt.Sprintf("%s-part%d%s", fileName, 1, fileExtension)
		tmpFileName2 := fmt.Sprintf("%s-part%d%s", fileName, 2, fileExtension)

		tmpFile1, _ := os.Open(tmpFileName1)
		tmpFile2, _ := os.Open(tmpFileName2)
		defer func(tmpFile1 *os.File) {
			err := tmpFile1.Close()
			if err != nil {
				fmt.Println("File close error")
			}
			err = os.Remove(tmpFileName1)
			if err != nil {
				fmt.Println("File Remove error")
			}
		}(tmpFile1)
		defer func(tmpFile2 *os.File) {
			err := tmpFile2.Close()
			if err != nil {
				fmt.Println("File close error")
			}
			err = os.Remove(tmpFileName2)
			if err != nil {
				fmt.Println("File Remove error")
			}
		}(tmpFile2)

		// 파일 내용 수신하여 저장
		reader1 := bufio.NewReader(tmpFile1)
		reader2 := bufio.NewReader(tmpFile2)
		var byteCnt int64
		var errorCnt int64
		for {
			if errorCnt == 2 {
				break
			}
			if byteCnt%2 == 0 {
				// get 1 byte from file
				b, err := reader1.ReadByte()
				if err != nil {
					errorCnt++
					continue // finish to reach file end
				}
				_, err = file.Write([]byte{b})
				if err != nil {
					return
				}
			} else {
				// get 1 byte from file
				b, err := reader2.ReadByte()
				if err != nil {
					errorCnt++
					continue // finish to reach file end
				}
				_, err = file.Write([]byte{b})
				if err != nil {
					return
				}
			}
			byteCnt++
		}

		fmt.Printf("%s file store sucessful!\n", fileName)
	} else { // otherwise case
		fmt.Println("Invalid argument.")
	}
}

// put file to server
func sendFile(fileName string, serverName string, serverPort string, part int) {

	// connect to TCP server
	conn, err := net.Dial("tcp", serverName+":"+serverPort)
	if err != nil {
		fmt.Println("fail to connect server: ", err)
		return
	}

	// prepare command string to send command
	commandStr := "put"
	conn.Write([]byte(commandStr))
	commandBuffer := make([]byte, 1024)
	read, _ := conn.Read(commandBuffer)
	// is server response is not ok
	if string(commandBuffer[:read]) != "ok" {
		fmt.Println("fail to receive fileName")
		os.Exit(1)
	}

	// try to open file
	originalFile, err := os.Open(fileName)
	if err != nil {
		return
	}

	// get file extension from fileName
	fileExtension := filepath.Ext(fileName)
	// get file name without extension from fileName
	fileName = fileName[0 : len(fileName)-len(fileExtension)]
	// join that strings
	fileName = fmt.Sprintf("%s-part%d%s", fileName, part+1, fileExtension)
	// send to target File name
	conn.Write([]byte(fileName))
	fileNameBuffer := make([]byte, 1024)
	read, _ = conn.Read(fileNameBuffer)
	if string(fileNameBuffer[:read]) != "ok" {
		fmt.Println("fail to receive command")
		os.Exit(1)
	}

	// try to get file Stats
	fileInfo, err := originalFile.Stat()
	if err != nil {
		fmt.Println("fail to get file stat:", err)
		return
	}

	// convert file Size string to Integer
	size := strconv.FormatInt(fileInfo.Size(), 10)
	// send file Size to server
	conn.Write([]byte(size))
	fileSizeBuffer := make([]byte, 1024)
	read, _ = conn.Read(fileSizeBuffer)
	if string(fileSizeBuffer[:read]) != "ok" {
		fmt.Println("fail to receive fileName")
		os.Exit(1)
	}

	// get Buffer from file
	reader := bufio.NewReader(originalFile)
	var byteCount int
	for {
		// get 1 byte from file
		b, err := reader.ReadByte()
		if err != nil {
			break // finish to reach file end
		}

		// check odd, even byte
		if byteCount%2 == part {
			_, err := conn.Write([]byte{b})
			if err != nil {
				return
			}
		}
		byteCount++
	}

	fmt.Printf("%s send successful\n", fileName)
	conn.Close()
}

func receiveFile(fileName string, serverName string, serverPort string, part int) {
	// connect to TCP server
	conn, err := net.Dial("tcp", serverName+":"+serverPort)
	if err != nil {
		fmt.Println("fail to connect server: ", err)
		return
	}

	// prepare command string to send command
	commandStr := "get"
	conn.Write([]byte(commandStr))
	commandBuffer := make([]byte, 1024)
	read, _ := conn.Read(commandBuffer)
	// is server response is not ok
	if string(commandBuffer[:read]) != "ok" {
		fmt.Println(string(commandBuffer[:read]))
		os.Exit(1)
	}

	// get file extension from fileName
	fileExtension := filepath.Ext(fileName)
	// get file name without extension from fileName
	fileName = fileName[0 : len(fileName)-len(fileExtension)]
	// join that strings
	fileName = fmt.Sprintf("%s-part%d%s", fileName, part+1, fileExtension)
	// request file Name
	conn.Write([]byte(fileName))

	// try to get fileSize
	fileSizeBuffer := make([]byte, 1024)
	read, _ = conn.Read(fileSizeBuffer)
	fileSize, err := strconv.ParseInt(string(fileSizeBuffer[:read]), 10, 64)
	if err != nil {
		fmt.Println("fail to transfer file size:", err)
		conn.Write([]byte("fail to transfer file size"))
		return
	} else {
		conn.Write([]byte("ok"))
	}
	// 파일 생성
	file, err := os.Create(fileName)
	if err != nil {
		fmt.Println("fail to create file:", err)
		return
	}
	// 파일 내용 수신하여 저장
	var receivedBytes int64
	buffer := make([]byte, 1024)
	for {
		n, err := conn.Read(buffer)
		fmt.Println(fileName)
		if err != nil {
			if err != io.EOF {
				fmt.Println("fail to read file:", err)
			}
			break
		}
		receivedBytes += int64(n)

		// 파일에 받은 내용 쓰기
		if _, err := file.Write(buffer[:n]); err != nil {
			fmt.Println("fail to write file:", err)
			return
		}

		// 파일이 전부 받아졌는지 확인
		if receivedBytes >= fileSize {
			file.Close()
			break
		}
	}

	fmt.Printf("%s file store sucessful!\n", fileName)
}

File: hw5/bouns_assignment/SplitFileClient.java
/*
SplitFileClient.java
20190532 sang yun lee
*/

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

class SplitFileClient {
	public static void main(String[] args) {
		// argument handling
		if (args.length != 2){
			System.out.println("Invalid argument.");
			System.exit(1);
		}
		// get argument from system args
		String commandName = args[0];
		String fileName = args[1];
		final String threadFileName = fileName;

		// server Info hardcoding
		String firstServerName = "nsl2.cau.ac.kr";
		String firstServerPort = "40532";
		String secondServerName = "nsl5.cau.ac.kr";
		String secondServerPort = "50532";

		ExecutorService executor = Executors.newFixedThreadPool(2);
		CountDownLatch latch = new CountDownLatch(2);
		// put command case
		if (commandName.equals("put")){
			Runnable task1 = () -> {
				try {
					sendFile(threadFileName, firstServerName, firstServerPort, 0);		
				} finally {
					latch.countDown();
				}
			};
			Runnable task2 = () -> {
				try {
					sendFile(threadFileName, secondServerName, secondServerPort, 1);
				} finally {
					latch.countDown();
				}
			};
			executor.submit(task1);
			executor.submit(task2);
			// Wait for all tasks to complete
            try {
                latch.await();  // Wait until the count reaches 0
                executor.shutdown();  // Shutdown the executor
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.out.println("Thread interrupted");
            }
		}// get command case
		else if (commandName.equals("get")){
			Runnable task1 = () -> {
				try {
					recieveFile(threadFileName, firstServerName, firstServerPort, 0);		
				} finally {
					latch.countDown();
				}
			};
			Runnable task2 = () -> {
				try {
					recieveFile(threadFileName, secondServerName, secondServerPort, 1);
				} finally {
					latch.countDown();
				}
			};
			executor.submit(task1);
			executor.submit(task2);
			// Wait for all tasks to complete
            try {
                latch.await();  // Wait until the count reaches 0
                executor.shutdown();  // Shutdown the executor
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                System.out.println("Thread interrupted");
            }

			int idx = fileName.lastIndexOf('.');
			String fileExtension = fileName.substring(idx);
			fileName = fileName.substring(0, idx);
			String outputFileName = String.format("%s-merged%s", fileName, fileExtension);
			String tmpFileName1 = String.format("%s-part%d%stmp%s", fileName, 1, fileExtension, fileExtension);
			String tmpFileName2 = String.format("%s-part%d%stmp%s", fileName, 2, fileExtension, fileExtension);
			File tmpFile1 = new File(tmpFileName1);
			File tmpFile2 = new File(tmpFileName2);
			try (
				FileOutputStream fos = new FileOutputStream(outputFileName);
				FileInputStream fis1 = new FileInputStream(tmpFile1);
				FileInputStream fis2 = new FileInputStream(tmpFile2)
			){
				long byteCnt = 0;
				int finishCnt = 0;
				while (true){
					if (finishCnt == 2)
						break;
					if (byteCnt % 2 == 0){
						byte[] b = new byte[1];
						if (fis1.read(b) <= 0){
							finishCnt++;
							continue;
						}
						fos.write(b);
					} else {
						byte[] b = new byte[1];
						if (fis2.read(b) <= 0){
							finishCnt++;
							continue;
						}
						fos.write(b);
					}
					byteCnt++;
				}
				System.out.println(threadFileName + "file merge successful!");
			} catch (Exception e) {
				System.out.println("File Open Error");
				System.exit(1);
			}
			if (tmpFile1.exists() && tmpFile1.delete()){
				System.out.println("delete Tempfile1 successful");
			} else {
				System.out.println("delete TempFile1 failed");
			}
			if (tmpFile2.exists() && tmpFile2.delete()){
				System.out.println("delete Tempfile2 successful");
			} else {
				System.out.println("delete TempFile2 failed");
			}
		}
	}

	// put file to server
	private static void sendFile(String fileName, String serverName, String serverPort, int part){
		// connet to TCP Server
		try (
			Socket socket = new Socket(serverName, Integer.parseInt(serverPort));
			OutputStream os = socket.getOutputStream();
			InputStream is = socket.getInputStream();
		) {
			int read;

			// prepare command string to send command
			String commandStr = "put";
			// send to server
			os.write(commandStr.getBytes());
			// get from server
			byte[] commandBuffer = new byte[1024];
			read = is.read(commandBuffer);
			String commandRes = new String(commandBuffer, 0, read);
			// server response is not ok
			if (!commandRes.equals("ok")){
				System.out.println("fail to receive fileName");
				throw new IOException();
			}
			System.out.println("Request to server to put :" + fileName);

			// get file Object
			File srcFile = new File(fileName);

			// extract fileExtension from fileName
			int idx = fileName.lastIndexOf('.');
			String fileExtension = fileName.substring(idx);
			// get fileName without extension from fileName
			fileName = fileName.substring(0, idx);
			// join that strings
			fileName = String.format("%s-part%d%s", fileName, part + 1, fileExtension);
			// send file Name to server
			os.write(fileName.getBytes());
			// get from response
			byte[] fileNameBuffer = new byte[1024];
			read = is.read(fileNameBuffer);
			String fileNameRes = new String(fileNameBuffer, 0, read);
			if (!fileNameRes.equals("ok")){
				System.out.println("fail to receive fileName");
				throw new IOException();
			}

			// try to get fileSize
			if (srcFile.exists() && srcFile.isFile()){
				// get file Size
				long size = srcFile.length();
				// send fileSize to server
				os.write(Long.toString(size).getBytes());
				// wait response from server
				byte[] fileSizeResBuffer = new byte[1024];
				read = is.read(fileSizeResBuffer);
				String fileSizeRes = new String(fileSizeResBuffer, 0, read);
				if (!fileSizeRes.equals("ok")){
					System.out.println("fail to receive fileSize");
					throw new IOException();
				}
			}

			// create fileInputStream object using Input data to file
			try (FileInputStream fis = new FileInputStream(srcFile)){
				int byteCount = 0;
				byte[] b = new byte[1];
				// get 1 byte from origin file
				while (fis.read(b) > 0){
					if (byteCount % 2 == part){
						os.write(b);
					}
					byteCount++;
				}
				System.out.println(fileName + " send successful");
			} catch (Exception e) {
				System.out.println("File Error");
				System.exit(1);
			}
		} catch(IOException e){
			System.out.println("fail to connect server");
			System.exit(1);
		}
	}

	private static void recieveFile(String fileName, String serverName, String serverPort, int part){
		System.out.println("Request to server to get : " + fileName);
		try (
			Socket socket = new Socket(serverName, Integer.parseInt(serverPort));
			InputStream is = socket.getInputStream();
			OutputStream os = socket.getOutputStream();
		) {
			int read;
			String commandStr = "get";
			// send to server
			os.write(commandStr.getBytes());
			// get from server
			byte[] commandResBuffer = new byte[1024];
			read = is.read(commandResBuffer);
			String commandRes = new String(commandResBuffer, 0, read);
			if (!commandRes.equals("ok")){
				System.out.println("fail to receive fileName");
				throw new IOException();
			}
			
			// extract fileName, fileExtension, formating
			int idx = fileName.lastIndexOf('.');
			String fileExtension = fileName.substring(idx);
			fileName = fileName.substring(0, idx);
			fileName = String.format("%s-part%d%s", fileName, part + 1, fileExtension);
			os.write(fileName.getBytes());
			// get file Size from server
			byte[] fileSizeBuffer = new byte[1024];
			read = is.read(fileSizeBuffer);
			String fileSizeString = new String(fileSizeBuffer, 0, read);
			long fileSize = Long.parseLong(fileSizeString);
			os.write("ok".getBytes());
			
			
			// get file Object
			File tmpFile = new File(fileName + "tmp" + fileExtension);
			try (FileOutputStream fos = new FileOutputStream(tmpFile)){
				long receivedBytes = 0;
				byte[] buffer = new byte[1024];
				int bytesRead;
				while (receivedBytes < fileSize && (bytesRead = is.read(buffer)) != -1) { 
					fos.write(buffer, 0, bytesRead);
					receivedBytes += bytesRead;	
				}
				System.out.println(fileName + " store successful");
			} catch (Exception e) {
				System.out.println("File Error");
				System.exit(1);
			}

		} catch(IOException e){
			System.out.println("fail to connect server");
			System.exit(1);
		}
	}
}
